SAM-142 - add docker compose for development

eapearson opened the pull request:
This set of changes brings a docker compose local development workflow. The docker compose configuration deploys the sample service and all required services (arangodb, kafka, a mock server for KBase workspace and auth) so that an operational sample service instance can be used and iterated against locally. There is currently no local development workflow involving a running server, as far as I know.

## Existing testing workflow

The existing local testing workflow requires running the sample service and associated services directly on the host, which has a higher burden on the developer to download and run the required binaries individually, as well as ensure the host meets the requirements for each of them (e.g. proper version of Java).

However, this workflow has not been adapted for having a local service running, which is typically a requirement for iterating on a service, and especially evaluating it in the context of implementing and iterating on api methods. 

## Motivation

The initial motivation was a need to add a couple of new methods in support of front end tools. I needed to be able to evaluate the utility and performance of the methods in the context of those tools. To this end, I would need to stand up a local instance, and direct the front end to use this sample service. This is an approach I've used often for KBase dynamic services, and occasionally for core services. Dynamic services work well in this context, because their service dependencies are nearly always KBase core services. It is easy to proxy requests to all KBase core services, because they work in a consistent manner, and are used through their public interfaces, requiring only a KBase login token for authorization. Core services are more difficult because they usually depend on private connections to sensitive resources, like databases, internal services, or the filesystem.

## Overview of services

The docker compose support provided in this PR attempts to solve this by running instances of the required services locally. 3rd party services run in their canonical docker containers, KBase services are mocked using a simple mock server.

### 3rd party services

- ArangoDB is run in a container using the official project image. After launching it is initialized with the collections necessary for the operation of the sample service. However, at this time, the database is not seeded. Any data would need to be added through the sample service api.

- Kafka is run in a container using the official project image. It requires no data seeding. 

- Zookeeper is also run in an official image; it is a dependency of Kafka.

### KBase Services

KBase services are mocked using a simple mocking server developed for local ui development. It is a server that we developed, and provides support for JSON-RPC 1.1, JSON-RPC 2.0, and REST services. It is in an unfinished state, I'd consider it alpha level, but does work well enough to be used in development and testing scenarios.

Each service requires data to be available in `dev/data/SampleService`. In this PR there is only a single sample data file. In future work this will be expanded. To be more specific, this PR was extracted from an effort to add two methods to the sample service; when that effort is continued, the docker compose support will be better exercised and expanded to support that development scenario.

## Status

The docker compose setup works well enough to start all dependent services, and the sample service. The sample service "status" endpoint responds correctly. Other endpoints have not been utilized directly. However, this work derives from a larger effort to utilize docker compose for testing. In that case, all integration and system tests utilize part or all of the docker compose services. As such, we know that it handles all api calls, including those to create samples and fetch samples, create, modify and query acls, etc.

That work will be available in a future PR, and will be based on this one.

## Usage

To run these containers:

- requirements: git, docker, make
- clone the repo
- enter there repo directory
- enter "make host-start-dev-server"
  - if on macOS Monterey "PORT=5001 make host-start-dev-server"

For further details see `docs/using-development-docker-compose.md`.

## Associated changes

This work also provides basic documentation on using Piping on macOS, running the docker compose configuration and  minor fixes and improvements to the Dockerfile.

## Future Work

- identify ArangoDB data required by the sample service but provided by the Relation Engine
- Scripts and data to create a fully functioning data set in ArangoDB, workspace, and auth; enough to exercise any api method
- Auto restarting of the sample service upon code file changes (supports local development iterative workflow).
- Adapt tests to utilize these containers
- Adapt GHA workflows to utilize those new tests
- Consolidation of development configuration data

scanon reviewed (changes requested):
This looks good.  Just one requested change.

eapearson commented on Dockerfile:
```diff
@@ -12,21 +12,14 @@ RUN \
 
-# really need a test build and a prod build. Not sure that's possible via sdk.
-# RUN pipenv install --system --deploy --ignore-pipfile --dev
+RUN mkdir -p /kb/module/work \
```
How true; update in place.

I'd also like to, not now, but sometime, replace the dockerize  with the semi-official fork https://github.com/powerman/dockerize. Steve had forked it to add support for KBase, but most of that has been merged into this fork, as well as other fixes, improvements - and it is supported.
The one-time downside of that switch is that there is a small difference in the command line options which can affect rancher setups, but in this case the dockerize is buried in the entrypoint file, so that is not an issue.

eapearson commented:
@scanon I've updated the Dockerfile, and synced from master, resolving a merge conflict. Would you mind having another look at it?

eapearson commented:
@scanon I added instructions for running the containers in the background with different scenarios (show logs, redirect logs to files, send to /dev/null), and simplified to just start/stop (combining remove with stop).

eapearson commented:
Added DETACH flag, which defaults to "no" (but that can be changed), to use the native docker compose detached mode. Added that to instructions.

scanon reviewed (approved):
LGTM
